### PR TITLE
refactor: tests/utils.go cleanup

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
`tests/utils.go` file contains several utility functions that can be considered as duplicated or that are only used once.
For the first ones. many things can be achieved with the already present `libvmi`/etc funcions;
for the second ones there are no reasons to store them in an utility file.
This PR aims to extract and refactor a small part of the tests/utils.go file.
In particular, in this PR we:
- unexport `TestCleanup` function.
- move `NewRandomMigration` function in the `libmigration` pkg.
- replace `AddEphemeralCdrom` function with `libvmi.WithEphemeralCDRom` option.
- move `AddBootOrderToDisk` to its only one usage.
- replace `AddPVCDisk`  function with `libvmi.WithPersistentVolumeClaim` option.
- replace `NewRandomVMIWithPVC` function with `libvmi.New` with proper options.
- replace `AddConfigMapDisk`, `AddSecretDisk`, `AddServiceAccountDisk`,
`AddLabelDownwardAPIVolume` and `AddWatchdog` functions with proper options.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] ~Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required~
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] ~Upgrade: Impact of this change on upgrade flows was considered and addressed if required~
- [ ] ~Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test~
- [ ] ~Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.~
- [ ] ~Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered~

**Release note**:

```release-note
NONE
```
/sig code-quality

### CODE REVIEW

File: tests/libmigration/migration.go
Comment (dankenigsberg) [inline tests/libmigration/migration.go:40]: Can we drop the "Random" bit? This function is very systematic.

Comment (dankenigsberg) [reply tests/libmigration/migration.go:40]: P.s. is this commit new? I don't recall it in former revisions of this PR. Since it touches so many files, it makes sense to move it to a new PR, so as not to block the merge of this. Also, can you add content to the commit message?

File: tests/tests_suite_test.go
Comment (dankenigsberg) [inline tests/tests_suite_test.go:145]: To me, it would be easier to review this change in its own PR.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/libmigration/migration.go
Comment (fossedihelm) [reply tests/libmigration/migration.go:40]: If we want to move it in a separate PR then I will drop the commit.

File: tests/tests_suite_test.go
Comment (fossedihelm) [reply tests/tests_suite_test.go:145]: This does not look a huge change IMHO. The commit is very easy to read and affect only 2 files, with few line changes. Am I missing smth? Thanks

File: tests/libmigration/migration.go
Comment (dankenigsberg) [reply tests/libmigration/migration.go:40]: doh

Comment (dankenigsberg) [reply tests/libmigration/migration.go:40]: > doh

If it wasn't clear, I indeed confused unrelated PRs. My recommendation to review this commit elsewhere stands, but it's your call. It's mainly about the probability of git rebase conflicts.

Comment (fossedihelm) [reply tests/libmigration/migration.go:40]: Thanks! Let's try to give it a try! If several `needs-rebase` will occur due to that change I will drop the commit. WDYT?

File: tests/migration/migration.go
Comment (0xFelix) [inline tests/migration/migration.go:514]: How was this handled before?

File: tests/migration/migration.go
Comment (0xFelix) [inline tests/migration/migration.go:1391]: Should we make these the default maybe? I see you need to add them in a lot of places.

File: tests/migration/migration.go
Comment (0xFelix) [inline tests/migration/migration.go:538]:
```suggestion
				vmi, err = virtClient.VirtualMachineInstance(testsuite.GetTestNamespace(vmi)).Create(context.Background(), vmi)
```

WDYT?

File: tests/migration/migration.go
Comment (0xFelix) [inline tests/migration/migration.go:525]:
```suggestion
				dv, err = virtClient.CdiClient().CdiV1beta1().DataVolumes(testsuite.GetTestNamespace(dv)).Create(context.Background(), dv, metav1.CreateOptions{})
```

File: tests/operator/operator.go
Comment (0xFelix) [inline tests/operator/operator.go:1936]: Use `testsuite.GetTestNamespace` here too?

File: tests/storage/snapshot.go
Comment (0xFelix) [inline tests/storage/snapshot.go:1344]: Put it in L1346?

File: tests/storage/snapshot.go
Comment (0xFelix) [inline tests/storage/snapshot.go:1343]: Duplicated in L1345

**Comment by fossedihelm** [comment]
Thanks @0xFelix for the review and the suggestion. I pushed the requested changes. PTAL

File: tests/migration/migration.go
Comment (fossedihelm) [reply tests/migration/migration.go:1391]: I'm not totally sure about it. In the migration file you see it in a lot of places because we want to create migratable    vmi. I don't know if it is worth it to create a `WithMigrationEnabled`(smth like this) option that wrap the two options. In any case I tend to consider it part of a separate PR, due to the, already big, size of the PR. WDYT?
Thanks

File: tests/libmigration/migration.go
Comment (dankenigsberg) [reply tests/libmigration/migration.go:40]: > Ok for the name!

care to consider a shorter name?

File: tests/migration/migration.go
Comment (fossedihelm) [reply tests/migration/migration.go:514]: This was handled in the same way in the `NewRandomVirtualMachineInstanceWithBlockDisk` function.
```
func NewRandomVirtualMachineInstanceWithBlockDisk(imageUrl, namespace string, accessMode k8sv1.PersistentVolumeAccessMode) (*v1.VirtualMachineInstance, *cdiv1.DataVolume) {
	if !libstorage.HasCDI() {
		Skip("Skip DataVolume tests when CDI is not present")
	}
	sc, exists := libstorage.GetRWOBlockStorageClass()
	if accessMode == k8sv1.ReadWriteMany {
		sc, exists = libstorage.GetRWXBlockStorageClass()
	}
	if !exists {
		Skip("Skip test when Block storage is not present")
	}
```

File: tests/libmigration/migration.go
Comment (fossedihelm) [reply tests/libmigration/migration.go]: Missed change xD
Done!
Thanks

File: tests/dryrun_test.go
Comment (fossedihelm) [reply tests/dryrun_test.go:235]: Are you referring to the `NewRandomMigration` to `NewMigration`?
If yes, Done :)
Otherwise, I agree that typically is a good idea to add a description! Here the things seemed to me quite self explanatory.
Please let me know, thanks

Comment (dankenigsberg) [reply tests/dryrun_test.go:235]: yes, I meant
```
refactor: rename NewRandomMigration to NewMigration

Remove the `Random` part in the name function, since
there's nothing random. The only thing that is (almost)
random is the name, that is generated, but this pattern
is used in almost every resource in the testsuite (see vmi).

```

File: tests/libmigration/migration.go
Comment (0xFelix) [inline tests/libmigration/migration.go:40]: What about calling it just `New`? The `libmigration` already explains what it is about. (similar to `libvmi.New()`)

File: tests/migration/migration.go
Comment (0xFelix) [reply tests/migration/migration.go:1391]: Yes, not as part of this PR. But we might should consider adding an option like that?

File: tests/migration/migration.go
Comment (0xFelix) [reply tests/migration/migration.go:514]: ACK

File: tests/migration/migration.go
Comment (dankenigsberg) [reply tests/migration/migration.go:1391]: I don't think that this would be necessary. I believe that @AlonaKaplan is working to make the default ("bridge") binding migratable. We should still test with Masquerade because it is so much better than bridge (no stealing IP address from virt-launcher)

File: tests/libmigration/migration.go
Comment (fossedihelm) [reply tests/libmigration/migration.go:40]: I like that! Done! Thanks

File: tests/dryrun_test.go
Comment (0xFelix) [inline tests/dryrun_test.go:28]: nit: newline

File: tests/operator/operator.go
Comment (0xFelix) [reply tests/operator/operator.go:40]: same nit again :)

**Comment by vladikr** [comment]
@fossedihelm thanks, looks fine to me. 
It just needs a rebase. 

/approve

File: tests/migration/migration.go
Comment (EdDev) [inline tests/migration/migration.go:466]: If you move the `libvmi.WithMasqueradeNetworking()...` at the end, will it work?

```go
vmi := libvmi.NewAlpineWithTestTooling(
    libvmi.WithEphemeralCDRom("cdrom-0", v1.DiskBusSATA, cd.ContainerDiskFor(cd.ContainerDiskAlpine)),
    libvmi.WithEphemeralCDRom("cdrom-1", v1.DiskBusSCSI, cd.ContainerDiskFor(cd.ContainerDiskAlpine)),
    libvmi.WithMasqueradeNetworking()...,
)
```

File: tests/vmi_configuration_test.go
Comment (EdDev) [inline tests/vmi_configuration_test.go:3064]: I think you do not need this anymore.

**Review by EdDev — COMMENTED** [review]
This PR is great!!

I had only small suggestions.

File: tests/migration/migration.go
Comment (EdDev) [reply tests/migration/migration.go:466]: Self answering... this does not work with Go. :/

File: tests/vmi_configuration_test.go
Comment (fossedihelm) [reply tests/vmi_configuration_test.go]: Done

**Review by EdDev — APPROVED** [review]

**Comment by brianmcarey** [comment]
/hold

@fossedihelm - it looks like the sig-operator lanes have never passed against this PR. 

Can you take a look?
https://prow.ci.kubevirt.io/pr-history/?org=kubevirt&repo=kubevirt&pr=11056

**Comment by fossedihelm** [comment]
> /hold
> 
> @fossedihelm - it looks like the sig-operator lanes have never passed against this PR.
> 
> Can you take a look? https://prow.ci.kubevirt.io/pr-history/?org=kubevirt&repo=kubevirt&pr=11056

Thank you @brianmcarey for noticing it. I missed it!
Missing masquerade interface for migration. Fixed!
/unhold

**Review by 0xFelix — APPROVED** [review]
/lgtm
/retest

**Review by 0xFelix — APPROVED** [review]
/lgtm

**Pull request merged**
